File: puppyscript/puppy/syntactic.py
import re

from .ids import with_cmd, break_cmd, firsts, stmt_add_brc, ret, WORD, FLAGS, word_last, spec_name_of, moveTo


class Syntactic:
    content: str

    def __init__(self, file: str):
        with open(file, "r", encoding="utf-8") as f:
            self.content = f.read()

    def scan_returns(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.match(rf"(\s*){ret}\s+(.+)", line)
            if tmp is None:
                result.append(line)
            else:
                result.append("%s %s" % (tmp.group(1) + with_cmd,
                                         stmt_add_brc(tmp.group(2))))
                result.append(tmp.group(1) + ret)
        self.content = "\n".join(result)

    def break_shortcuts(self):
        self.content = re.sub(
            r"break(\s*\n)", break_cmd + r" false\1", self.content)
        self.content = re.sub(
            r"conti(\s*\n)", break_cmd + r" true\1", self.content)

    def unmap_import(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.fullmatch(r"(\s*)(use|unmap)\s+import\s+(.+)", line)
            if tmp is None:
                result.append(line)
            else:
                result.append(tmp.group(1) + "import " + tmp.group(3))
                result.append(tmp.group(1) + "unmap")
        self.content = "\n".join(result)

    def scan_import(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.fullmatch(rf"(\s*)import\s+({WORD})\s+as\s+({WORD})", line)
            if tmp is None:
                result.append(line)
            else:
                result.append("%s %s" % (tmp.group(1) + "import", tmp.group(2)))
                result.append("%s %s" % (tmp.group(1) + moveTo, tmp.group(3)))
        self.content = "\n".join(result)

    def use_statements(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.match(rf"(\s*)use\s*({WORD})(\s+as\s+({WORD}))?\s*", line)
            if tmp is None:
                result.append(line)
            else:
                if tmp.group(4) is None:
                    result.append(tmp.group(1) + word_last(tmp.group(2)) + " = " + tmp.group(2))
                else:
                    result.append(tmp.group(1) + tmp.group(4) + " = " + tmp.group(2))
        self.content = "\n".join(result)

    def special_functions(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.fullmatch(rf"(\s*)spec\s+{FLAGS}(func|proc)\s+({WORD})\s*:\s*", line)
            if tmp is not None:
                line = tmp.group(1) + \
                       tmp.group(3) + " " + spec_name_of(tmp.group(4)) + ":"
            result.append(line)
        self.content = "\n".join(result)

    def array_indexes(self):
        result = list()
        success = False
        for line in self.content.split("\n"):
            tmp = re.search(rf"\b({WORD})\[({WORD})]", line)
            if tmp is not None:
                success = True
                line = line[:tmp.start(1)] + \
                       "(%s.at_ref %s).data" % (tmp.group(1), tmp.group(2)) + \
                       line[tmp.end(2) + 1:]
            result.append(line)
        self.content = "\n".join(result)
        return success

    def all_array_indexes(self):
        while self.array_indexes():
            ...

    def scan_link_to(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.fullmatch(rf"(\s*)link\s+({WORD})\s+to\s+({WORD})\s*", line)
            if tmp is None:
                result.append(line)
            else:
                result.append(tmp.group(1) + tmp.group(3) + "." + word_last( tmp.group(2)) + " = " + tmp.group(2))
        self.content = "\n".join(result)

    def scan_extern(self):
        result = list()
        indent = -1
        externals = list()
        succeeded = False
        pub_all = False
        for line in self.content.split("\n"):
            if indent != -1:
                cur_indent = len(firsts(line))
                if cur_indent < indent:
                    indent = -1
                    externals.clear()
                    pub_all = False
                elif pub_all or cur_indent == indent:
                    for name in externals:
                        line = re.sub(r"([^\w.&])%s(\b)" % name, r"\1&%s\2" % name, line)
                result.append(line)
            else:
                tmp = re.match(r"(\s*)pub(_all)?\s+(.*)", line)
                if tmp is None:
                    result.append(line)
                else:
                    indent = tmp.group(1)
                    if tmp.group(2):
                        pub_all = True
                    for name in tmp.group(3).split(" "):
                        if len(name) > 0 and not name.isspace():
                            externals.append(name)
                    indent = len(indent)
                    succeeded = True
        self.content = "\n".join(result)
        return succeeded

    def scan_all_extern(self):
        self.content = re.sub(rf" ", r"\1pub \2\1\2 = \3", self.content)
        while self.scan_extern():
            ...

    def work(self, output_name: str) -> str:
        self.scan_returns()
        self.break_shortcuts()
        self.unmap_import()
        self.scan_import()
        self.use_statements()
        self.special_functions()
        self.all_array_indexes()
        self.scan_link_to()
        self.scan_all_extern()
        output = output_name + ".syn"
        with open(output, "w", encoding="utf-8") as f:
            f.write(self.content)
        return output

File: puppyscript/cpp-puppy/syntactic.cpp
//
// Created by impodog on 7/30/2023.
//
#include "processor.h"

std::regex RE_RETURN(R"((\s*)ret\s+(.+?)\s*)"),
        RE_BREAK(R"((\s*)break\s*)"),
        RE_CONTINUE(R"((\s*)conti\s*)"),
        RE_UNMAP_IMPORT(R"((\s*)(use|unmap)\s+import\s+(.+))"),
        RE_IMPORT_AS(R"((\s*)import\s+()" WORD R"()\s+as\s+()" WORD R"())"),
        RE_USE_AS(R"((\s*)use\s+()" WORD R"()\s+as\s+()" WORD R"())"),
        RE_SPECIAL_FUNCTION("(\\s*)spec\\s+" FLAGS "(func|proc)\\s+(" WORD ")\\s*:\\s*"),
        RE_LINK_TO("(\\s*)link\\s+(" WORD ")\\s+to\\s+(" WORD ")\\s*"),
        RE_EXTERN(R"((\s*)pub(_all)?\s+(.*))"),
        RE_PUB_ASSIGN("(\\s*)pub\\s+(" WORD ")\\s*=\\s*(.+)");

class Syntactic {
    std::string content;

    void scan_returns() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_RETURN)) {
                result.push_back(add_with(indent_of(line), results[2].str()));
                result.push_back(indent_of(line) + "ret");
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    void scan_break_and_continue() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            if (std::regex_match(line, RE_BREAK)) {
                result.push_back(indent_of(line) + "break false");
            } else if (std::regex_match(line, RE_CONTINUE)) {
                result.push_back(indent_of(line) + "break true");
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    void scan_unmap_import() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_UNMAP_IMPORT)) {
                result.push_back(results[1].str() + "import " + results[3].str());
                result.push_back(results[1].str() + "unmap");
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    void scan_import_as() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_IMPORT_AS)) {
                result.push_back(results[1].str() + "import " + results[2].str());
                result.push_back(results[1].str() + "mov " + results[3].str());
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    void scan_use_as() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_USE_AS)) {
                result.push_back(add_with(results[1].str(), results[2].str()));
                if (results[4].matched) {
                    result.push_back(results[1].str() + "mov " + results[3].str());
                } else {
                    result.push_back(add_with(results[1].str(), results[2].str()));
                    result.push_back(results[1].str() + "mov " + word_base(results[2].str()));
                }
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    void scan_special_function() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_SPECIAL_FUNCTION)) {
                result.push_back(results[1].str() + results[3].str()
                                 + " " + add_special(results[4].str()) + ":");
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    void scan_link_to() {
        string_list result;
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_LINK_TO)) {
                result.push_back(results[1].str() +
                                 results[3].str() + "." + word_base(results[2].str())
                                 + " = " + results[2].str());
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    bool scan_extern() {
        string_list result;
        size_t indent = -1;
        string_list externals;
        bool succeeded = false;
        bool pub_all = false;
        auto lines = split(content, '\n');
        for (auto line: lines) {
            if (indent != -1) {
                auto cur_indent = indent_of(line).size();
                if (cur_indent < indent) {
                    indent = -1;
                    externals.clear();
                    pub_all = false;
                } else if (pub_all || cur_indent == indent) {
                    for (auto &name: externals) {
                        line = std::regex_replace(line, std::regex("([^\\w.&])" + name + "(\\b)"),
                                                  "$1&" + name + "$2");
                    }
                }
                result.push_back(line);
            } else {
                std::smatch results;
                if (std::regex_match(line, results, RE_EXTERN)) {
                    indent = results[1].str().size();
                    if (results[2].matched) {
                        pub_all = true;
                    }
                    externals = split(results[3].str(), ' ');
                    succeeded = true;
                } else {
                    result.push_back(line);
                }
            }
        }
        content = join(result, '\n');
        return succeeded;
    }

    void scan_all_extern() {
        content = std::regex_replace(content, RE_PUB_ASSIGN, R"($1pub $2$1$2 = $3)");
        while (scan_extern());
    }

public:
    explicit Syntactic(const path &path) {
        content = read(path);
    }

    path work(const std::string &output_name) {
        scan_returns();
        scan_break_and_continue();
        scan_unmap_import();
        scan_import_as();
        scan_use_as();
        scan_special_function();
        scan_link_to();
        scan_all_extern();

        path output = output_name + ".syntactic.puppy";
        write(output, content);
        return output;
    }

};

PROCESSOR(syntactic) {
    Syntactic syntactic(file);
    return syntactic.work(output_name);
}